### PR TITLE
CY-1060 Add plugin installation handling, and filtering & merging logic

### PR DESCRIPTION
**Pull request by davidginzbourg**

### CODE REVIEW

File: cloudify/plugins/workflows.py [hunk 1]
```diff
@@ -746,6 +749,15 @@ def _reinstall():
                                            related_nodes=intact_nodes,
                                            ignore_failure=ignore_failure)
 
+    def _handle_plugins():
```
Comment (tyacbovi) [inline cloudify/plugins/workflows.py]: I think _update_plugins is more readable

File: cloudify/plugins/workflows.py [hunk 2]
```diff
@@ -746,6 +749,15 @@ def _reinstall():
...
+        """Uninstalls plugins that need to be uninstalled and then installs
```
Comment (tyacbovi) [inline cloudify/plugins/workflows.py]: I think the func is clear enough so that doc string is not needed

File: cloudify/tests/test_utils.py [hunk 3]
```diff
@@ -20,11 +20,13 @@
 import unittest
 
 from cloudify.exceptions import CommandExecutionException
+from dsl_parser.constants import PLUGIN_INSTALL_KEY, PLUGIN_NAME_KEY
```
Comment (tyacbovi) [inline cloudify/tests/test_utils.py]: style - Different libs are in different sections

File: cloudify/tests/test_utils.py [hunk 4]
```diff
@@ -65,3 +67,35 @@ def test_executable_no_override(self):
     @mock.patch.dict(os.environ, {'CFY_EXEC_TEMP': '/fake/temp'})
     def test_executable_override(self):
         self.assertEqual('/fake/temp', get_exec_tempdir())
+
+
+class TestPluginFunctions(unittest.TestCase):
```
Comment (tyacbovi) [inline cloudify/tests/test_utils.py]: what about the other funcs ?

File: cloudify/utils.py [hunk 5]
```diff
@@ -36,6 +36,7 @@
 from cloudify.state import workflow_ctx, ctx
 from cloudify.constants import SUPPORTED_ARCHIVE_TYPES
 from cloudify.amqp_client import BlockingRequestResponseHandler
+from dsl_parser.constants import PLUGIN_INSTALL_KEY, PLUGIN_NAME_KEY
```
Comment (tyacbovi) [inline cloudify/utils.py]: the same style comment

File: cloudify/utils.py [hunk 6]
```diff
@@ -647,3 +648,88 @@ def get_admin_api_token():
 
 
 internal = Internal()
+
+
+def extract_plugins_to_install(plugin_list, filter_func):
+    """
```
Comment (tyacbovi) [inline cloudify/utils.py]: The docstring is too similar to the impl, it's needs to describe what part it's doing

File: cloudify/utils.py [hunk 7]
```diff
@@ -647,3 +648,88 @@ def get_admin_api_token():
...
+    wf_plugins = extract_plugins_to_install(workflow_plugins, filter_func)
+    return dep_plugins + wf_plugins
+
+
+def add_plugins_to_install(ctx, plugins_to_install, sequence):
+    """Adds a task to the sequence that installs the plugins.
+    """
+    if plugins_to_install:
+        sequence.add(
+            ctx.send_event('Installing deployment plugins'),
```
Comment (tyacbovi) [inline cloudify/utils.py]: What about wf plugins?

File: cloudify/utils.py [hunk 8]
```diff
@@ -647,3 +648,88 @@ def get_admin_api_token():
...
+            ctx.execute_task('cloudify_agent.operations.install_plugins',
+                             kwargs={'plugins': plugins_to_install}))
+
+
+def add_plugins_to_uninstall(ctx, plugins_to_uninstall, sequence):
+    """Adds a task to the sequence that uninstalls the plugins.
+    """
+    if plugins_to_uninstall:
+        sequence.add(
+            ctx.send_event('Uninstalling deployment plugins'),
```
Comment (tyacbovi) [inline cloudify/utils.py]: What about wf plugins?

File: cloudify/tests/test_utils.py (hunk 4 above)
Comment (davidginzbourg) [reply cloudify/tests/test_utils.py]: Thought they're too simple. I'll add them too.

File: cloudify/utils.py (hunk 6 above)
Comment (davidginzbourg) [reply cloudify/utils.py]: Changed

File: cloudify/utils.py (hunk 8 above)
Comment (davidginzbourg) [reply cloudify/utils.py]: Added it to the event log message.

File: cloudify/utils.py (hunk 7 above)
Comment (davidginzbourg) [reply cloudify/utils.py]: Added it to the event log message.

File: cloudify/tests/test_utils.py [hunk 9]
```diff
@@ -65,3 +68,99 @@ def test_executable_no_override(self):
...
+        sequence_mock.add.assert_called_with('send_event', 'execute_task')
+
+    def test_add_plugin_to_uninstall_doesnt_add_to_sequence(self):
+        ctx_mock = mock.Mock()
+        sequence_mock = mock.Mock()
+        _test_task_not_added_to_sequence(
+            sequence_mock, ctx_mock, utils.add_plugins_to_uninstall)
+
+
+def _test_task_not_added_to_sequence(sequence_mock, ctx_mock, add_func):
```
Comment (tyacbovi) [inline cloudify/tests/test_utils.py]: maybe use 'verify_task_not_called'

File: cloudify/tests/test_utils.py [hunk 10]
```diff
@@ -65,3 +68,99 @@ def test_executable_no_override(self):
...
+        ctx_mock.send_event.assert_called()
+        ctx_mock.execute_task.assert_called_with(
+            task_name='cloudify_agent.operations.uninstall_plugins',
+            kwargs={
+                'plugins': plugins_list,
+                'delete_managed_plugins': False})
+        sequence_mock.add.assert_called_with('send_event', 'execute_task')
+
+    def test_add_plugin_to_uninstall_doesnt_add_to_sequence(self):
+        ctx_mock = mock.Mock()
```
Comment (tyacbovi) [inline cloudify/tests/test_utils.py]: ctx_mock and sequence_mock are not needed to be created always

File: cloudify/tests/test_utils.py [hunk 11]
```diff
@@ -65,3 +68,99 @@ def test_executable_no_override(self):
...
+            dep_plugins, wf_plugins, lambda x: x['dummy'] == 2)
+        self.assertListEqual(dep_plugins[1:] + wf_plugins, res)
+
+    def test_add_plugin_to_install_adds_to_sequence(self):
+        ctx_mock = mock.Mock()
+        ctx_mock.send_event.return_value = 'send_event'
+        ctx_mock.execute_task.return_value = 'execute_task'
+        sequence_mock = mock.Mock()
+        plugins_list = ['something']
+        utils.add_plugins_to_install(ctx_mock, plugins_list, sequence_mock)
```
Comment (tyacbovi) [inline cloudify/tests/test_utils.py]: maybe divide to several test classes

**Pull request merged**